File: scripts/release-metadata.mjs
import { createHash } from "node:crypto";
import { execFileSync } from "node:child_process";
import {
  createReadStream,
  existsSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import path from "node:path";
import { pathToFileURL } from "node:url";

const SEMVER = /^[0-9]+\.[0-9]+\.[0-9]+$/;
const GIT_REVISION = /^(?:[a-f0-9]{40}|[a-f0-9]{64})$/;
const SHA256 = /^[a-f0-9]{64}$/;
const IMAGE_ID = /^sha256:[a-f0-9]{64}$/;
const IMAGE_NAMES = ["engine", "frontend", "postgres"];
const SOURCE_STATES = new Set(["clean", "dirty", "unavailable"]);
const IMAGE_MODES = new Set(["built", "prebuilt"]);

export function inspectGitSource(root, version) {
  try {
    const revision = git(root, ["rev-parse", "--verify", "HEAD"]).trim().toLowerCase();
    if (!GIT_REVISION.test(revision)) throw new Error("unsupported Git object format");
    const status = git(root, ["status", "--porcelain", "--untracked-files=normal"]);
    const tag = matchingAnnotatedTag(root, revision, version);
    return { revision, state: status.trim() ? "dirty" : "clean", tag };
  } catch {
    return { revision: null, state: "unavailable", tag: null };
  }
}

export async function sha256File(file) {
  const hash = createHash("sha256");
  const stream = createReadStream(file);
  for await (const chunk of stream) hash.update(chunk);
  return hash.digest("hex");
}

export async function createReleaseMetadata(options) {
  validateBuildOptions(options);
  const source = options.source ?? inspectGitSource(options.projectRoot, options.version);
  validateSource(source, options.version);
  const references = {
    engine: options.engineImage,
    frontend: options.frontendImage,
    postgres: options.postgresImage,
  };
  const contentIds = options.imageIdentities ?? inspectDockerImageIdentities(references);
  validateImageIdentities(contentIds);
  const archiveHash = await sha256File(options.imageArchive);

  return {
    schemaVersion: 1,
    package: {
      name: options.packageName,
      version: options.version,
      createdAt: timestampToIso(options.packageTimestamp),
    },
    source,
    compose: {
      file: "docker-compose.yml",
      source: normalizeRelativePath(options.composeSource, "compose source"),
    },
    images: {
      mode: options.imageMode,
      references,
      contentIds,
      archive: {
        file: path.basename(options.imageArchive),
        sha256: archiveHash,
      },
    },
    integrity: {
      algorithm: "sha256",
      manifest: "checksums.sha256",
    },
  };
}

export function inspectDockerImageIdentities(references) {
  const identities = {};
  for (const name of IMAGE_NAMES) {
    const reference = references[name];
    let identity;
    try {
      identity = execFileSync("docker", ["image", "inspect", "--format", "{{.Id}}", reference], {
        encoding: "utf8",
        stdio: ["ignore", "pipe", "pipe"],
      }).trim().toLowerCase();
    } catch (error) {
      const detail = error?.stderr?.toString().trim() || error?.message || String(error);
      throw new Error(`cannot inspect ${name} image ${reference}: ${detail}`);
    }
    if (!IMAGE_ID.test(identity)) {
      throw new Error(`${name} image ${reference} returned an invalid content ID`);
    }
    identities[name] = identity;
  }
  return identities;
}

export async function writeReleaseMetadata(options) {
  const metadata = await createReleaseMetadata(options);
  const output = path.resolve(options.output);
  const temporary = `${output}.tmp-${process.pid}`;
  try {
    writeFileSync(temporary, `${JSON.stringify(metadata, null, 2)}\n`, { mode: 0o644 });
    renameSync(temporary, output);
  } finally {
    rmSync(temporary, { force: true });
  }
  return metadata;
}

export async function verifyPackageReleaseMetadata(packageDirectory, expectations = {}) {
  const root = path.resolve(packageDirectory);
  const metadataFile = path.join(root, "release-metadata.json");
  const metadata = parseJsonFile(metadataFile);
  validateMetadata(metadata);
  validateReleaseExpectations(metadata, expectations);
  if (metadata.package.name !== path.basename(root)) {
    throw new Error(`package directory is ${path.basename(root)}; metadata names ${metadata.package.name}`);
  }

  const archiveFile = safePackageFile(root, metadata.images.archive.file, "image archive");
  const actualArchiveHash = await sha256File(archiveFile);
  if (actualArchiveHash !== metadata.images.archive.sha256) {
    throw new Error("image archive SHA-256 does not match release metadata");
  }
  safePackageFile(root, metadata.compose.file, "compose file");

  const checksumFile = safePackageFile(root, metadata.integrity.manifest, "checksum manifest");
  const checksums = parseChecksums(readFileSync(checksumFile, "utf8"));
  for (const file of ["release-metadata.json", metadata.images.archive.file]) {
    const expected = checksums.get(file);
    if (!expected) throw new Error(`checksum manifest does not include ${file}`);
    const actual = await sha256File(path.join(root, file));
    if (actual !== expected) throw new Error(`checksum manifest does not match ${file}`);
  }
  return metadata;
}

export function validateReleaseExpectations(metadata, expectations = {}) {
  if (!isRecord(expectations)) throw new Error("release expectations must be an object");
  const errors = [];
  if (expectations.version !== undefined) {
    if (typeof expectations.version !== "string" || !SEMVER.test(expectations.version)) {
      throw new Error("expected version must use x.y.z");
    }
    compareExpected(errors, "package version", metadata.package.version, expectations.version);
  }
  if (expectations.revision !== undefined) {
    if (typeof expectations.revision !== "string") throw new Error("expected source revision is invalid");
    const revision = expectations.revision.toLowerCase();
    if (!GIT_REVISION.test(revision)) throw new Error("expected source revision is invalid");
    compareExpected(errors, "source revision", metadata.source.revision, revision);
  }
  if (expectations.tag !== undefined) {
    if (
      typeof expectations.tag !== "string"
      || !/^v[0-9]+\.[0-9]+\.[0-9]+$/.test(expectations.tag)
    ) {
      throw new Error("expected source tag must use vx.y.z");
    }
    compareExpected(errors, "source tag", metadata.source.tag, expectations.tag);
  }
  if (expectations.sourceState !== undefined) {
    if (!SOURCE_STATES.has(expectations.sourceState)) throw new Error("expected source state is invalid");
    compareExpected(errors, "source state", metadata.source.state, expectations.sourceState);
  }
  if (expectations.imageMode !== undefined) {
    if (!IMAGE_MODES.has(expectations.imageMode)) throw new Error("expected image mode is invalid");
    compareExpected(errors, "image mode", metadata.images.mode, expectations.imageMode);
  }
  if (errors.length > 0) throw new Error(errors.join("\n"));
}

function validateBuildOptions(options) {
  if (!options || typeof options !== "object") throw new Error("release metadata options are required");
  if (!/^[A-Za-z0-9][A-Za-z0-9._-]*$/.test(options.packageName ?? "")) {
    throw new Error("package name must be a portable file name");
  }
  if (!SEMVER.test(options.version ?? "")) throw new Error("version must use x.y.z");
  timestampToIso(options.packageTimestamp);
  normalizeRelativePath(options.composeSource, "compose source");
  if (!IMAGE_MODES.has(options.imageMode)) throw new Error("image mode must be built or prebuilt");
  for (const [label, value] of Object.entries({
    "engine image": options.engineImage,
    "frontend image": options.frontendImage,
    "PostgreSQL image": options.postgresImage,
  })) {
    if (typeof value !== "string" || !value.trim() || /[\r\n]/.test(value)) {
      throw new Error(`${label} must be a non-empty single-line reference`);
    }
  }
  if (typeof options.projectRoot !== "string" || !options.projectRoot) {
    throw new Error("project root is required");
  }
  if (typeof options.imageArchive !== "string" || !existsSync(options.imageArchive)) {
    throw new Error("image archive does not exist");
  }
}

function validateMetadata(metadata) {
  if (!isRecord(metadata) || metadata.schemaVersion !== 1) {
    throw new Error("release metadata schemaVersion must be 1");
  }
  if (!isRecord(metadata.package) || !SEMVER.test(metadata.package.version ?? "")) {
    throw new Error("release metadata package version is invalid");
  }
  if (!/^[A-Za-z0-9][A-Za-z0-9._-]*$/.test(metadata.package.name ?? "")) {
    throw new Error("release metadata package name is invalid");
  }
  if (
    typeof metadata.package.createdAt !== "string"
    || !/^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/.test(metadata.package.createdAt)
    || Number.isNaN(Date.parse(metadata.package.createdAt))
  ) {
    throw new Error("release metadata creation time is invalid");
  }
  validateSource(metadata.source, metadata.package.version);
  if (!isRecord(metadata.compose)) throw new Error("release metadata compose section is missing");
  if (metadata.compose.file !== "docker-compose.yml") {
    throw new Error("release metadata compose file is invalid");
  }
  normalizeRelativePath(metadata.compose.file, "compose file");
  normalizeRelativePath(metadata.compose.source, "compose source");
  if (!isRecord(metadata.images) || !IMAGE_MODES.has(metadata.images.mode)) {
    throw new Error("release metadata image mode is invalid");
  }
  if (!isRecord(metadata.images.references)) throw new Error("image references are missing");
  for (const name of IMAGE_NAMES) {
    const value = metadata.images.references[name];
    if (typeof value !== "string" || !value || /[\r\n]/.test(value)) {
      throw new Error(`release metadata contains an invalid ${name} image reference`);
    }
  }
  validateImageIdentities(metadata.images.contentIds);
  if (!isRecord(metadata.images.archive) || !SHA256.test(metadata.images.archive.sha256 ?? "")) {
    throw new Error("release metadata image archive is invalid");
  }
  normalizeRelativePath(metadata.images.archive.file, "image archive");
  if (metadata.images.archive.file !== "cyanrex-images.tar") {
    throw new Error("release metadata image archive file is invalid");
  }
  if (
    !isRecord(metadata.integrity)
    || metadata.integrity.algorithm !== "sha256"
    || metadata.integrity.manifest !== "checksums.sha256"
  ) {
    throw new Error("release metadata integrity section is invalid");
  }
}

function validateImageIdentities(identities) {
  if (!isRecord(identities)) throw new Error("image content IDs are missing");
  for (const name of IMAGE_NAMES) {
    if (!IMAGE_ID.test(identities[name] ?? "")) {
      throw new Error(`release metadata contains an invalid ${name} image content ID`);
    }
  }
}

function validateSource(source, version) {
  if (!isRecord(source) || !SOURCE_STATES.has(source.state)) {
    throw new Error("source state must be clean, dirty, or unavailable");
  }
  if (source.state === "unavailable") {
    if (source.revision !== null || source.tag !== null) {
      throw new Error("unavailable source must not claim a revision or tag");
    }
    return;
  }
  if (!GIT_REVISION.test(source.revision ?? "")) throw new Error("source revision is invalid");
  if (source.tag !== null && source.tag !== `v${version}`) {
    throw new Error(`source tag must be v${version} or null`);
  }
}

function matchingAnnotatedTag(root, revision, version) {
  const reference = `refs/tags/v${version}`;
  try {
    if (git(root, ["cat-file", "-t", reference]).trim() !== "tag") return null;
    const target = git(root, ["rev-list", "-n", "1", reference]).trim().toLowerCase();
    return target === revision ? `v${version}` : null;
  } catch {
    return null;
  }
}

function timestampToIso(value) {
  const match = /^(\d{4})(\d{2})(\d{2})-(\d{2})(\d{2})(\d{2})$/.exec(value ?? "");
  if (!match) throw new Error("package timestamp must use YYYYMMDD-HHMMSS UTC format");
  const iso = `${match[1]}-${match[2]}-${match[3]}T${match[4]}:${match[5]}:${match[6]}Z`;
  const parsed = new Date(iso);
  if (Number.isNaN(parsed.valueOf()) || parsed.toISOString().replace(".000Z", "Z") !== iso) {
    throw new Error("package timestamp is not a valid UTC time");
  }
  return iso;
}

function normalizeRelativePath(value, label) {
  if (typeof value !== "string" || !value || path.isAbsolute(value)) {
    throw new Error(`${label} must be a relative path`);
  }
  const normalized = value.replaceAll("\\", "/");
  if (normalized.split("/").includes("..")) throw new Error(`${label} must stay within the package`);
  return normalized.replace(/^\.\//, "");
}

function safePackageFile(root, file, label) {
  const normalized = normalizeRelativePath(file, label);
  const destination = path.resolve(root, normalized);
  if (path.dirname(destination) !== root || !existsSync(destination)) {
    throw new Error(`${label} is missing or outside the package root`);
  }
  return destination;
}

function parseChecksums(source) {
  const entries = new Map();
  for (const line of source.split("\n")) {
    if (!line.trim()) continue;
    const match = /^([a-f0-9]{64})\s+[ *]?([^\r\n]+)$/.exec(line);
    if (!match) throw new Error(`invalid checksum manifest line: ${line}`);
    if (entries.has(match[2])) throw new Error(`duplicate checksum entry for ${match[2]}`);
    entries.set(match[2], match[1]);
  }
  return entries;
}

function parseJsonFile(file) {
  try {
    return JSON.parse(readFileSync(file, "utf8"));
  } catch (error) {
    throw new Error(`cannot read ${file}: ${error instanceof Error ? error.message : error}`);
  }
}

function isRecord(value) {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function compareExpected(errors, label, actual, expected) {
  if (actual !== expected) errors.push(`${label} is ${actual ?? "null"}; expected ${expected}`);
}

function git(root, arguments_) {
  return execFileSync("git", arguments_, {
    cwd: root,
    encoding: "utf8",
    stdio: ["ignore", "pipe", "pipe"],
  });
}

export function parseReleaseMetadataArguments(argv) {
  if (argv.length === 0 || argv.includes("-h") || argv.includes("--help")) return { help: true };
  if (argv[0] === "--verify") {
    if (!argv[1] || argv[1].startsWith("--")) {
      throw new Error("--verify requires one package directory before expectation options");
    }
    const values = parseNamedValues(argv.slice(2));
    const allowed = [
      "expect-version",
      "expect-revision",
      "expect-tag",
      "expect-source-state",
      "expect-image-mode",
    ];
    const unknown = Object.keys(values).filter((name) => !allowed.includes(name));
    if (unknown.length > 0) throw new Error(`unknown verify option --${unknown[0]}`);
    return {
      mode: "verify",
      packageDirectory: argv[1],
      expectations: {
        ...(values["expect-version"] ? { version: values["expect-version"] } : {}),
        ...(values["expect-revision"] ? { revision: values["expect-revision"] } : {}),
        ...(values["expect-tag"] ? { tag: values["expect-tag"] } : {}),
        ...(values["expect-source-state"] ? { sourceState: values["expect-source-state"] } : {}),
        ...(values["expect-image-mode"] ? { imageMode: values["expect-image-mode"] } : {}),
      },
    };
  }
  const values = parseNamedValues(argv);
  const required = [
    "output",
    "project-root",
    "package-name",
    "version",
    "package-timestamp",
    "engine-image",
    "frontend-image",
    "postgres-image",
    "image-mode",
    "compose-source",
    "image-archive",
  ];
  for (const name of required) {
    if (!(name in values)) throw new Error(`--${name} is required`);
  }
  const unknown = Object.keys(values).filter((name) => !required.includes(name));
  if (unknown.length > 0) throw new Error(`unknown option --${unknown[0]}`);
  return {
    mode: "write",
    output: values.output,
    projectRoot: values["project-root"],
    packageName: values["package-name"],
    version: values.version,
    packageTimestamp: values["package-timestamp"],
    engineImage: values["engine-image"],
    frontendImage: values["frontend-image"],
    postgresImage: values["postgres-image"],
    imageMode: values["image-mode"],
    composeSource: values["compose-source"],
    imageArchive: values["image-archive"],
  };
}

function parseNamedValues(argv) {
  const values = {};
  for (let index = 0; index < argv.length; index += 2) {
    const name = argv[index];
    const value = argv[index + 1];
    if (!name?.startsWith("--") || value === undefined) {
      throw new Error("metadata options require name/value pairs");
    }
    const key = name.slice(2);
    if (key in values) throw new Error(`duplicate option --${key}`);
    values[key] = value;
  }
  return values;
}

function printHelp() {
  console.log(`Usage:
  node scripts/release-metadata.mjs --output <file> --project-root <dir> --package-name <name> \\
    --version <x.y.z> --package-timestamp <YYYYMMDD-HHMMSS> --engine-image <image> \\
    --frontend-image <image> --postgres-image <image> --image-mode <built|prebuilt> \\
    --compose-source <relative-path> --image-archive <file>
  node scripts/release-metadata.mjs --verify <extracted-package-directory> \\
    [--expect-version <x.y.z>] [--expect-revision <commit>] [--expect-tag <vx.y.z>] \\
    [--expect-source-state <clean|dirty|unavailable>] [--expect-image-mode <built|prebuilt>]`);
}

async function main() {
  const options = parseReleaseMetadataArguments(process.argv.slice(2));
  if (options.help) return printHelp();
  if (options.mode === "verify") {
    const metadata = await verifyPackageReleaseMetadata(options.packageDirectory, options.expectations);
    console.log(`Release metadata verified for ${metadata.package.name}.`);
    return;
  }
  const metadata = await writeReleaseMetadata(options);
  console.log(`Release metadata created for ${metadata.package.name}.`);
}

const invokedAsScript = process.argv[1]
  && pathToFileURL(path.resolve(process.argv[1])).href === import.meta.url;
if (invokedAsScript) {
  main().catch((error) => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}

File: engine/src/bin/cyanrex-release/release_metadata.rs
use crate::release_archive::{decimal_triplet, lowercase_hex, parse_archive_name, portable_path};
use crate::strict_json;
use chrono::{DateTime, NaiveDateTime, Utc};
use serde_json::{Map, Value};
use std::collections::{BTreeMap, BTreeSet};

const IMAGE_NAMES: [&str; 3] = ["engine", "frontend", "postgres"];

pub struct Expectations {
    pub version: Option<String>,
    pub revision: Option<String>,
    pub tag: Option<String>,
    pub source_state: Option<String>,
    pub image_mode: Option<String>,
}

pub fn validate_metadata(
    source: &[u8],
    package_root: &str,
    archive_name: &str,
    hashes: &BTreeMap<String, String>,
) -> Result<Value, String> {
    let metadata_value = strict_json::from_slice(source)
        .map_err(|error| format!("cannot parse release metadata: {error}"))?;
    let metadata = object(&metadata_value, "release metadata")?;
    exact_keys(
        metadata,
        &[
            "schemaVersion",
            "package",
            "source",
            "compose",
            "images",
            "integrity",
        ],
        "release metadata",
    )?;
    if metadata["schemaVersion"].as_u64() != Some(1) {
        return Err("release metadata schemaVersion must be 1".to_owned());
    }
    let archive = parse_archive_name(archive_name)?;
    let package = object(&metadata["package"], "release metadata package")?;
    exact_keys(
        package,
        &["name", "version", "createdAt"],
        "release metadata package",
    )?;
    if package["name"].as_str() != Some(package_root) {
        return Err("release metadata package name does not match the archive".to_owned());
    }
    if !package["name"].as_str().is_some_and(valid_package_name) {
        return Err("release metadata package name is invalid".to_owned());
    }
    let version = package["version"].as_str().unwrap_or_default();
    if !decimal_triplet(version) {
        return Err("release metadata package version is invalid".to_owned());
    }
    if version != archive.version {
        return Err("release metadata package version does not match the archive".to_owned());
    }
    let created_at = parse_timestamp(&package["createdAt"], "release metadata package createdAt")?;
    if created_at != archive.timestamp {
        return Err("release metadata creation time does not match the archive name".to_owned());
    }

    validate_source(&metadata["source"], version)?;
    validate_compose(&metadata["compose"])?;
    validate_images(&metadata["images"], hashes)?;
    let integrity = object(&metadata["integrity"], "release metadata integrity")?;
    exact_keys(
        integrity,
        &["algorithm", "manifest"],
        "release metadata integrity",
    )?;
    if integrity["algorithm"].as_str() != Some("sha256")
        || integrity["manifest"].as_str() != Some("checksums.sha256")
    {
        return Err("release metadata integrity declaration is invalid".to_owned());
    }
    Ok(metadata_value)
}

pub fn enforce_expectations(metadata: &Value, expected: &Expectations) -> Result<(), String> {
    let package = object(&metadata["package"], "release metadata package")?;
    let source = object(&metadata["source"], "release metadata source")?;
    let images = object(&metadata["images"], "release metadata images")?;
    if let Some(version) = expected.version.as_deref() {
        if !decimal_triplet(version) {
            return Err("expected package version must use x.y.z".to_owned());
        }
        if package["version"].as_str() != Some(version) {
            return Err("candidate package version does not match expectation".to_owned());
        }
    }
    if let Some(revision) = expected.revision.as_deref() {
        let revision = revision.to_ascii_lowercase();
        if !lowercase_hex(&revision, &[40, 64]) {
            return Err("expected source revision is invalid".to_owned());
        }
        if source["revision"].as_str() != Some(&revision) {
            return Err("candidate source revision does not match expectation".to_owned());
        }
    }
    if let Some(tag) = expected.tag.as_deref() {
        if !tag.strip_prefix('v').is_some_and(decimal_triplet) {
            return Err("expected source Tag must use vx.y.z".to_owned());
        }
        if source["tag"].as_str() != Some(tag) {
            return Err("candidate source Tag does not match expectation".to_owned());
        }
    }
    if expected
        .source_state
        .as_deref()
        .is_some_and(|state| source["state"].as_str() != Some(state))
    {
        return Err("candidate source state does not match expectation".to_owned());
    }
    if expected
        .image_mode
        .as_deref()
        .is_some_and(|mode| images["mode"].as_str() != Some(mode))
    {
        return Err("candidate image mode does not match expectation".to_owned());
    }
    Ok(())
}

fn validate_source(value: &Value, version: &str) -> Result<(), String> {
    let source = object(value, "release metadata source")?;
    exact_keys(
        source,
        &["revision", "state", "tag"],
        "release metadata source",
    )?;
    let state = source["state"].as_str().unwrap_or_default();
    if !matches!(state, "clean" | "dirty" | "unavailable") {
        return Err("release metadata source state is invalid".to_owned());
    }
    if state == "unavailable" {
        if !source["revision"].is_null() || !source["tag"].is_null() {
            return Err("unavailable release source must not claim a revision or Tag".to_owned());
        }
    } else {
        if !source["revision"]
            .as_str()
            .is_some_and(|revision| lowercase_hex(revision, &[40, 64]))
        {
            return Err("release metadata source revision is invalid".to_owned());
        }
        let expected_tag = format!("v{version}");
        if !source["tag"].is_null() && source["tag"].as_str() != Some(&expected_tag) {
            return Err("release metadata source Tag does not match its version".to_owned());
        }
    }
    Ok(())
}

fn validate_compose(value: &Value) -> Result<(), String> {
    let compose = object(value, "release metadata compose")?;
    exact_keys(compose, &["file", "source"], "release metadata compose")?;
    if compose["file"].as_str() != Some("docker-compose.yml") {
        return Err("release metadata compose file is invalid".to_owned());
    }
    portable_path(&compose["source"], "release metadata compose source")?;
    Ok(())
}

fn validate_images(value: &Value, hashes: &BTreeMap<String, String>) -> Result<(), String> {
    let images = object(value, "release metadata images")?;
    exact_keys(
        images,
        &["mode", "references", "contentIds", "archive"],
        "release metadata images",
    )?;
    if !matches!(images["mode"].as_str(), Some("built" | "prebuilt")) {
        return Err("release metadata image mode is invalid".to_owned());
    }
    let references = object(&images["references"], "release metadata image references")?;
    let identities = object(&images["contentIds"], "release metadata image content IDs")?;
    exact_keys(
        references,
        &IMAGE_NAMES,
        "release metadata image references",
    )?;
    exact_keys(
        identities,
        &IMAGE_NAMES,
        "release metadata image content IDs",
    )?;
    for name in IMAGE_NAMES {
        nonempty_line(
            &references[name],
            &format!("release metadata {name} image reference"),
        )?;
        if !identities[name]
            .as_str()
            .and_then(|value| value.strip_prefix("sha256:"))
            .is_some_and(|hash| lowercase_hex(hash, &[64]))
        {
            return Err(format!(
                "release metadata {name} image content ID is invalid"
            ));
        }
    }
    let image_archive = object(&images["archive"], "release metadata image archive")?;
    exact_keys(
        image_archive,
        &["file", "sha256"],
        "release metadata image archive",
    )?;
    if image_archive["file"].as_str() != Some("cyanrex-images.tar") {
        return Err("release metadata image archive file is invalid".to_owned());
    }
    let archive_hash = image_archive["sha256"].as_str().unwrap_or_default();
    if !lowercase_hex(archive_hash, &[64]) {
        return Err("release metadata image archive SHA-256 is invalid".to_owned());
    }
    for required in ["docker-compose.yml", "cyanrex-images.tar"] {
        if !hashes.contains_key(required) {
            return Err(format!(
                "release metadata references missing package file {required}"
            ));
        }
    }
    if hashes.get("cyanrex-images.tar").map(String::as_str) != Some(archive_hash) {
        return Err("image archive SHA-256 does not match release metadata".to_owned());
    }
    Ok(())
}

fn parse_timestamp(value: &Value, label: &str) -> Result<DateTime<Utc>, String> {
    let text = value.as_str().unwrap_or_default();
    NaiveDateTime::parse_from_str(text, "%Y-%m-%dT%H:%M:%SZ")
        .map(|timestamp| timestamp.and_utc())
        .map_err(|_| format!("{label} must be an RFC3339 UTC timestamp"))
}

fn exact_keys(value: &Map<String, Value>, keys: &[&str], label: &str) -> Result<(), String> {
    let expected = keys
        .iter()
        .map(|key| (*key).to_owned())
        .collect::<BTreeSet<_>>();
    let actual = value.keys().cloned().collect::<BTreeSet<_>>();
    if expected == actual {
        return Ok(());
    }
    let mut details = Vec::new();
    let missing = expected.difference(&actual).cloned().collect::<Vec<_>>();
    let unexpected = actual.difference(&expected).cloned().collect::<Vec<_>>();
    if !missing.is_empty() {
        details.push(format!("missing {}", missing.join(", ")));
    }
    if !unexpected.is_empty() {
        details.push(format!("unexpected {}", unexpected.join(", ")));
    }
    Err(format!(
        "{label} fields are invalid: {}",
        details.join("; ")
    ))
}

fn object<'a>(value: &'a Value, label: &str) -> Result<&'a Map<String, Value>, String> {
    value
        .as_object()
        .ok_or_else(|| format!("{label} must be a JSON object"))
}

fn nonempty_line(value: &Value, label: &str) -> Result<(), String> {
    let text = value.as_str().unwrap_or_default();
    if text.is_empty() || text.contains(['\n', '\r']) {
        Err(format!("{label} must be a non-empty single-line string"))
    } else {
        Ok(())
    }
}

fn valid_package_name(value: &str) -> bool {
    let bytes = value.as_bytes();
    bytes.first().is_some_and(u8::is_ascii_alphanumeric)
        && bytes
            .iter()
            .all(|byte| byte.is_ascii_alphanumeric() || matches!(byte, b'.' | b'_' | b'-'))
}
